src/config/src/index.ts

```diff
@@ -0,0 +1,340 @@
+// Import Node.js Dependencies
+import path from "node:path";
+import fs from "node:fs";
+import { EventEmitter } from "node:events";
+import { isDeepStrictEqual } from "node:util";
+
+// Import Third-party Dependencies
+import Ajv, { JSONSchemaType } from "ajv";
+import Observable from "zen-observable";
+import * as TOML from "smol-toml";
+
+// Import Internal Dependencies
+import * as utils from "./utils.js";
+
+// CONSTANTS
+const kPayloadIdentifier = Symbol("payload");
+const kSchemaIdentifier = Symbol("schema");
+const kAjv = new Ajv({ useDefaults: true });
+const kDefaultExtension = ".json";
+const kSupportedExtensions = new Set([".json", ".toml"]);
+const kDefaultSchema = {
+  title: "Default config",
+  type: "object",
+  additionalProperties: true
+};
+
+export interface ConfigOptions<T> {
+  createOnNoEntry?: boolean;
+  autoReload?: boolean;
+  writeOnSet?: boolean;
+  jsonSchema?: JSONSchemaType<T>;
+}
```

Comment:
I'll personally add an option to inject FS to simplify testing

Reply:
Mmh, the only test (i think) that would benefit ioc is this one

```ts
it("should recreate file on read when file does not exists and createOnNoEntry is true", async() => {
  const configPath = path.join(__dirname, "fixtures", ".doesNotExists3");
  assert(fs.existsSync(configPath) === false);

  const config = new Config(configPath, { createOnNoEntry: true });
  await config.read({ boo: "boom" });
  await once(config, "configWritten");

  assert(fs.existsSync(configPath) === true);
  assert.deepEqual(config.payload, { boo: "boom" });

  fs.rmSync(configPath);
  assert(fs.existsSync(configPath) === false);

  await config.read();
  await once(config, "configWritten");
  assert(fs.existsSync(configPath) === true);
  assert.deepEqual(config.payload, { boo: "boom" });

  await config.close();
  fs.rmSync(configPath);
});
```

Dealing with a fs mock would essentially be the same thing/complexity, wouldn't it?

Reply:
Could be useful outside of the project (like in RC or other projects)